File: 0312-burst-balloons/0312-burst-balloons.cpp
class Solution {
public:
    int maxCoins(vector<int>& nums) {
        int n = nums.size();
        vector<int> arr(n+2);
        arr[0]=1;
        arr[n+1]=1;
        for(int i =0;i<n;i++){
            arr[i+1]=nums[i];
        }
        vector<vector<int>> dp (n+2,vector<int>(n+2,0));

        for(int len =2 ; len<n+2;len++){
            for(int left=0;left+len<n+2;left++){
                int right = left+len;
                for(int i =left+1;i<right;i++){
                    dp[left][right]= max(dp[left][right], arr[left]*arr[i]*arr[right] +dp[left][i]+ dp[i][right]);
                }
            }
        }

        return dp[0][n+1];
        
    }
};

File: 0312-burst-balloons/0312-burst-balloons.java
class Solution {
    int [][]dp;
    int []arr;
    private int solve(int left,int right){
        if(left+1 == right){
            return 0;
        }

        if(dp[left][right]!=-1){
            return dp[left][right];
        }
        int maxCoins=0;
        for(int k=left+1; k<right;k++){
            int coins = arr[left]*arr[k]*arr[right];
            maxCoins = Math.max(maxCoins, coins + solve(left,k) + solve(k, right));

        }

        return dp[left][right]=maxCoins;
    }
    public int maxCoins(int[] nums) {
        int n = nums.length;
        arr = new int [n+2];
        dp = new int[n+2][n+2];
        arr[0]=1;
        arr[n+1]=1;
        for(int i =0;i<nums.length;i++){
            arr[i+1]=nums[i];
        }
        for(int [] row : dp ){
            Arrays.fill(row,-1);
        }
        return solve(0,n+1);
        
    }
}